stan/math/prim/fun/grad_2F1.hpp

```diff
@@ -8,10 +8,60 @@
 #include <stan/math/prim/fun/fabs.hpp>
 #include <stan/math/prim/fun/inv.hpp>
 #include <stan/math/prim/fun/log.hpp>
+#include <stan/math/prim/fun/sign.hpp>
 #include <cmath>
 
 namespace stan {
 namespace math {
+namespace internal {
+/**
+ * Returns the intermediate calculations for the gradient of the
+ * hypergeometric function (2F1).
+ *
+ * @tparam T scalar type or `var_value`
+ * @tparam T_array array of doubles or `var_values`
+ * @tparam T_array_int array of int
+ * @tparam T_int integer
+ * @param[out] log_g_old log of previous gradient values
+ * @param[out] log_g_old_sign sign of previous gradient values
+ * @param[out] log_t_new log of power series calculation
+ * @param[out] log_t_new_sign sign of power series calcuation
+ * @param[in] p intermediate power series value
+ * @param[in] log_z log_z see hypergeometric function 2F1 definition
+ * @param[in] log_t_old log of previous power series calculation
+ * @param[in] log_t_old_sign log of previous power series calculation
+ * @param[in] k loop iteration
+ * @param[in] a1 a1 see hypergeometric function 2F1 definition
+ * @param[in] a2 a2 see hypergeometric function 2F1 definition
+ * @param[in] b1 b1 see hypergeometric function 2F1 definition
+ */
+template <typename T, typename T_array, typename T_array_int, typename T_int>
```

Comment:
`std::array` and stan arrays are two very different things. I like the having more verbose names but I'd call these like `StdVec` and `StdVecInt`. Also note we use CamelCase for template parameter names.